STCON-121: Allow 'headers' to be a function and interpreted

AndrewIsh opened the pull request:
It is conceivable that a request may contain headers that need value
substitution to take place (as per query parameters, path components
etc.). This commit adds that.

The ability for the 'headers' value to be a function as well as an
object, as per 'path' etc. This allows us to access props that have
been passed to the component.

The use case this was implemented for was to allow an external
API key, that was being passed down to a component as a prop, to be used
in an authentication header for an API call being made by the child
component.

MikeTaylor commented on RESTResource/RESTResource.js:
```diff
@@ -324,6 +324,17 @@ export default class RESTResource {
         options.params = options.params(parsedQuery, _.get(props, ['match', 'params']), mockProps(state, this.module, props.dataKey, this.logger).resources, this.logger, props);
       }
 
+      // headers
```
This block of code looks fine.

My only hesitation is whether it would be better to abstract out the repeated block into a function that can be invoked to do this for both `params` and `headers` — also opening the way for it to be trivial to do the same for other parts of the `options` object.

AndrewIsh replied:
Good point, I've now updated the PR accordingly.

MikeTaylor replied:
Nice!

MikeTaylor commented on test/RESTResource.js:
```diff
@@ -43,6 +44,11 @@ describe('RESTResource', () => {
         .should.equal('/whatever/water/anyways');
     });
 
+    it('replaces header values', () => {
```
This test is fine so far as it goes, but it's a bit weak. Can you easily test the automatic expansion of `headers`?

AndrewIsh replied:
I agree @MikeTaylor, it **is** weak! I've spent a while trying to figure out what a better test would look like and have, so far, failed. The current tests only test `substitute` and the changes I made are in `verbOptions`. I think I'm not 100% clear on what you mean by "automatic expansion of headers", sorry if I'm being dim!

I intend creating tests for my new `buildOption` method as it needs some and Sonarcloud is complaining.

MikeTaylor replied:
OK, well you shouldn't worry too much about that. Testing is a means to an end.

AndrewIsh replied:
Thanks @skomorokh You're absolutely right, the test naming was completely inaccurate! I've not modified it accordingly and moved it into its own commit. Thanks for the feedback.

MikeTaylor reviewed (changes requested):
Apart from the comments above, the main things missing are:
* an entry in the changelog
* documentation tweaks in `doc/api.md`

axelhunn commented on RESTResource/RESTResource.js:
```diff
@@ -324,6 +324,17 @@ export default class RESTResource {
+          header => substitute(header, props, state, this.module, this.logger, this.dataKey)
+        );
+      } else if (typeof options.headers === 'function') {
+        const parsedQuery = queryString.parse(_.get(props, ['location', 'search']));
+        options.headers = options.headers(parsedQuery, _.get(props, ['match', 'params']), mockProps(state, this.module, props.dataKey, this.logger).resources, this.logger, props);
```
```suggestion
        const parsedQuery = queryString.parse(props?.location?.search);
        options.headers = options.headers(parsedQuery, props?.match?.params, mockProps(state, this.module, props.dataKey, this.logger).resources, this.logger, props);
```

MikeTaylor replied:
I think there are a lot of uses of `_.get`, predating the availability of the `?.` operator. I' not sure we necessarily want to fix these piecemeal.

doytch replied:
I'm all in favour of fixing things piecemeal. The fact is that nobody is gonna go, sit down, and remove all our reliance on lodash in a module in one swoop. So it's either we take our marginal gains here and there, or do nothing.

AndrewIsh replied:
As part of refactoring the building of the option object for `params` and `headers` into its own function, I took the opportunity to switch from `_.get` to optional chaining, it seemed a logic time to do it.

MikeTaylor replied:
Agree, given that this is now code that replaces the old inline version, it does make sense.

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint .

/home/jenkins/workspace/folio-org_stripes-connect_PR-173/project/RESTResource/RESTResource.js
  296:7   error  Assignment to function parameter 'option'  no-param-reassign
  307:7   error  Assignment to function parameter 'option'  no-param-reassign
  307:56  error  Multiple spaces found before 'mockProps'   no-multi-spaces

✖ 3 problems (3 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

axelhunn commented on RESTResource/RESTResource.js:
```diff
@@ -289,6 +289,27 @@ export default class RESTResource {
+        option,
+        param => substitute(param, props, state, this.module, this.logger, this.dataKey)
+      );
+      for (const key of Object.keys(option)) {
+        if (option[key] === null) {
+          return null;
+        }
+      }
```
```suggestion
      const someParamIsNull = Object.values(option).some(value => value === null)
      if (someParamIsNull) return null; // or toReturn = null

```
Is that the point here, return `null` if some of the params is null?

AndrewIsh replied:
> Is that the point here, return null if some of the params is null?

That seems to be the intent of that portion of code, however, it isn't a result of my refactor, it was just part of the pre-existing code that was refactored.

I'm not sure allowing the scope of my PR to creep by making arbitrary changes is a good idea, I'd rather keep the focus on the intention of my PR. I implemented the suggestion to switch from _.get to optional chaining following the comments from @MikeTaylor and @doytch since a move to deprecate _.get seems inevitable, but your change, although perfectly valid, seems like it's moving outside of the scope of the PR.

Happy to be shouted down though! :)

axelhunn replied:
I didn't mean to shout down )
In fact, I'm more concerned about not the syntax but the API - returning `null` if one of the params is exactly `null` seems strange, out of context.

doytch replied:
From what I can see, both the implementation in the current commit and the implementation that @axelhunn is suggesting are functionally equivalent. The only difference I see is short-circuiting the loop via `.some` instead of always running over every iteration of the object. The performance is not relevant here, but I think Alex's version does make it a bit clearer.

I totally understand any reticence to change the guts of the algorithm, but in this case I'd say you're well within the scope of the PR.

And yeah, the intention is to return `null` if anything computes to `null`. I think `null` is consciously selected here over undefined to attempt hammer home the point that the resource's owner purposely halted this fetch because it cannot be completed properly.

AndrewIsh replied:
@axelhunn Sorry, I didn't mean that you were shouting me down! I meant that I felt it was out of scope to make the change you were suggesting, but I was happy to do it if it was felt that the change should be included :+1:

As far as your question as to the sanity of returning null in the event that any of the params or headers is null, that's a good question.  It certainly seems to be intentional since @MikeTaylor [refactored it](https://github.com/folio-org/stripes-connect/commit/d8f91b1afd0d66884a61abd6fee1d6aef9aca40a)  at one point to keep Jenkins happy, and the [original commit](https://github.com/folio-org/stripes-connect/commit/31485ee969d1ccec4a2c01ed7962e177fd0a43d4) gives a bit of context.

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint .

/home/jenkins/workspace/folio-org_stripes-connect_PR-173/project/RESTResource/RESTResource.js
  301:82  error  Missing semicolon                        semi
  304:5   error  Block must not be padded by blank lines  padded-blocks

✖ 2 problems (2 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint .

/home/jenkins/workspace/folio-org_stripes-connect_PR-173/project/RESTResource/RESTResource.js
  263:3   error  Trailing spaces not allowed              no-trailing-spaces
  271:3   error  Trailing spaces not allowed              no-trailing-spaces
  282:80  error  Missing semicolon                        semi
  285:3   error  Block must not be padded by blank lines  padded-blocks

/home/jenkins/workspace/folio-org_stripes-connect_PR-173/project/test/RESTResource.js
  86:46  warning  'resources' is defined but never used. Allowed unused args must match /^_/u  no-unused-vars
  86:57  warning  'logger' is defined but never used. Allowed unused args must match /^_/u     no-unused-vars
  86:65  warning  'props' is defined but never used. Allowed unused args must match /^_/u      no-unused-vars
  86:65  error    'props' is already declared in the upper scope                               no-shadow

✖ 8 problems (5 errors, 3 warnings)
  4 errors and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

AndrewIsh commented:
I **think** this is now addressing all the feedback. Thanks so much for all the excellent input, I'm very grateful. I'm not sure why SonarCloud is complaining about coverage on the new `buildOption` function, it clearly has tests, though I'm not experienced with SonarCloud, so I could be misinterpreting its output.

Anyway, as I say, I think I've now done everything that was suggested. What's the next step? :+1:

doytch commented:
> I **think** this is now addressing all the feedback. Thanks so much for all the excellent input, I'm very grateful. I'm not sure why SonarCloud is complaining about coverage on the new `buildOption` function, it clearly has tests, though I'm not experienced with SonarCloud, so I could be misinterpreting its output.

Sonar recently got changed to calculate its coverage based on RTL/Jest tests so it's fair to ignore it rn in places that are still using BigTest.

> Anyway, as I say, I think I've now done everything that was suggested. What's the next step? 👍

I agree, this looks good! Ooooone last thing since @zburke was just complaining about this on the stripes-arch call: to save him time when releasing everything, we ask that people create a JIRA ticket (it can be pretty barebones, this is for tracking mainly!) in the appropriate project (STCON in this case) for tracking. Tack the JIRA number (eg, STCON-123) onto the end of the change log entry, at the start of the PR title, and then when we're merging, we'll **Squash and merge** everything into one commit (for easier cherry-picking) rather than having a bunch of smaller commits.

AndrewIsh commented:
> [...] we ask that people create a JIRA ticket (it can be pretty barebones, this is for tracking mainly!) in the appropriate project (STCON in this case) for tracking.

Will do, I've not delved into FOLIO's Jira before, so this should be fun!

AndrewIsh commented:
I've now created the Jira ticket ([STCON-121](https://issues.folio.org/browse/STCON-121)) and renamed this PR accordingly. I've also added the ticket number to the CHANGELOG entry. I didn't seem to be able to assign myself to the ticket (I'm assuming perms), so could someone who has sufficient perms could do that please?

I _think_ that's everything! :)

MikeTaylor commented:
I've assigned the Jira issue to you — odd that you couldn't do it.

This is great work, and I think very much ready to go. Thank you, and congratulations!

AndrewIsh commented:
> I've assigned the Jira issue to you — odd that you couldn't do it.
> 
> This is great work, and I think very much ready to go. Thank you, and congratulations!

Thanks @MikeTaylor - I'm very grateful for all the help and advice you've given in getting this done!

MikeTaylor commented:
Delighted to help. And especially delighted that within five minutes of joining the project, you've hoisted yourself up to "core Stripes developer" level :-)

skomorokh reviewed (changes requested):
Apologies, just closing some tabs and I see I was oblivious to the "pending" label on these comments. So, um, a little late...

ViktorSoroka07 commented on RESTResource/RESTResource.js:
```diff
@@ -256,6 +256,38 @@ export function substitute(original, props, state, module, logger, dataKey) {
+ * @return object
+ */
+
+export function buildOption(option, props, state, module, logger, dataKey) {
+  let toReturn;
+  if (typeof option === 'object') {
+    toReturn = _.mapValues(
+      option,
+      param => substitute(param, props, state, module, logger, dataKey)
+    );
+    const someParamIsNull = Object.values(option).some(value => value === null);
+    if (someParamIsNull) return null; // or toReturn = null
+  } else if (typeof option === 'function') {
+    const parsedQuery = queryString.parse(props.location?.search);
+    toReturn = option(parsedQuery, props.match?.params, mockProps(state, module, dataKey, logger).resources, logger, props);
+  }
+  return toReturn;
+}
```
I would avoid if else, as early return can be used. Also `someParamIsNull` check can be made before computing the option.
```suggestion
export function buildOption(option, props, state, module, logger, dataKey) {
  if (typeof option === 'object') {
    const someParamIsNull = Object.values(option).some(value => value === null);
    
    if (someParamIsNull) return null;

    return _.mapValues(
      option,
      param => substitute(param, props, state, module, logger, dataKey)
    );
  }

   if (typeof option === 'function') {
    const parsedQuery = queryString.parse(props.location?.search);
    return option(parsedQuery, props.match?.params, mockProps(state, module, dataKey, logger).resources, logger, props);
  }

  return null;
}
```

AndrewIsh replied:
Thanks for the suggestion @ViktorSoroka07. I think this is pretty subjective to be honest. In my mind the if/else is more explicit than the early returns. I think your suggestion is perfectly valid, I just don't really know if it gives us anything, other than a different approach :)

ViktorSoroka07 replied:
So for me is early return is better Andrew, but this is really the style of preference:). So the only useful moment here still 
```
    toReturn = _.mapValues(
      option,
      param => substitute(param, props, state, module, logger, dataKey)
    );
    const someParamIsNull = Object.values(option).some(value => value === null);
    if (someParamIsNull) return null; // or toReturn = null
```
to move the `if (someParamIsNull) return null; // or toReturn = null` check before the `toReturn`.

AndrewIsh replied:
Thanks for that @ViktorSoroka07 I'm very grateful for your feedback. I've now moved the check for null values and the subsequent return to above the formation of `toReturn`, you're right, it removes the unnecessary formation of `toReturn`.

AndrewIsh commented on RESTResource/RESTResource.js:
```diff
@@ -256,6 +256,39 @@ export function substitute(original, props, state, module, logger, dataKey) {
+  if (typeof option === 'object') {
+    // If any of the option properties have null values, we can't go any further
+    const someParamIsNull = Object.values(option).some(value => value === null);
+    if (someParamIsNull) return null; // or toReturn = null
```
Ah yes, good spot, I'm not sure how I managed to leave that in there. Thanks!
